Samples/NetworkMediaStreamer/RTSPServer/src/SocketWrapper.cpp

```diff
@@ -0,0 +1,304 @@
+
+        winrt::check_hresult(QueryContextAttributes(
+            &m_hCtxt,
+            SECPKG_ATTR_STREAM_SIZES,
+            &m_secPkgContextStrmSizes));
+        auto newBufsz = m_secPkgContextStrmSizes.cbHeader + m_secPkgContextStrmSizes.cbMaximumMessage + m_secPkgContextStrmSizes.cbTrailer;
+        if (newBufsz > m_bufSz)
+        {
+            m_bufSz = newBufsz;
+            m_pInBuf.reset(std::make_unique<BYTE[]>(m_bufSz).release());
+            m_pOutBuf.reset(std::make_unique<BYTE[]>(m_bufSz).release());
+        }
+
+        // if TLS authentication fails, the rtsp session will ask for username/password digest auth
+        m_bIsAuthenticated = AuthenticateClient();
+
+    }
+}
+CSocketWrapper::~CSocketWrapper()
+{
+    //TODO: need to handle secure socket shutdown message to client
+}
+
+void CSocketWrapper::ReadDelegate(PVOID arg, BOOLEAN flag)
+{
+    auto pSock = (CSocketWrapper*)arg;
+    pSock->m_SecurityStatus  = SEC_E_INVALID_HANDLE;
+    try 
+    {
+        WSAResetEvent(pSock->m_readEvent.get());
+
+        auto numRead = recv(pSock->m_socket, (char*)pSock->m_pInBuf.get(), pSock->m_bufSz, 0);
+        if (numRead > 0)
+        {
+            TimeStamp         tokenLifetime;
+            SecBufferDesc     OutBuffDesc;
+            SecBuffer         OutSecBuff;
+            SecBufferDesc     InBuffDesc;
+            SecBuffer         InSecBuff[2];
+            ULONG             Attribs = ASC_REQ_MUTUAL_AUTH;
+
+            OutBuffDesc.ulVersion = 0;
+            OutBuffDesc.cBuffers = 1;
+            OutBuffDesc.pBuffers = &OutSecBuff;
+
+            OutSecBuff.cbBuffer = pSock->m_bufSz;
+            OutSecBuff.BufferType = SECBUFFER_TOKEN;
+            OutSecBuff.pvBuffer = pSock->m_pOutBuf.get();
+
+            InBuffDesc.ulVersion = 0;
+            InBuffDesc.cBuffers = 2;
+            InBuffDesc.pBuffers = InSecBuff;
+
+            InSecBuff[0].cbBuffer = numRead;
+            InSecBuff[0].BufferType = SECBUFFER_TOKEN;
+            InSecBuff[0].pvBuffer = pSock->m_pInBuf.get();
+
+            InSecBuff[1].cbBuffer = 0;
+            InSecBuff[1].BufferType = SECBUFFER_EMPTY;
+            InSecBuff[1].pvBuffer = nullptr;
+            bool bFirstHandshake = !(pSock->m_hCtxt.dwLower || pSock->m_hCtxt.dwUpper);
+            pSock->m_SecurityStatus = AcceptSecurityContext(
+                &pSock->m_hCred,
+                bFirstHandshake ? NULL : &pSock->m_hCtxt,
+                &InBuffDesc,
+                Attribs,
+                SECURITY_NETWORK_DREP,
+                bFirstHandshake ? &pSock->m_hCtxt : NULL,
+                &OutBuffDesc,
+                &Attribs,
+                &tokenLifetime);
+
+            if (InSecBuff[1].BufferType == SECBUFFER_EXTRA)
+            {
+                //TODO: handle this case
+            }
+
+            winrt::check_hresult(pSock->m_SecurityStatus);
+            send(pSock->m_socket, (char*)OutSecBuff.pvBuffer, OutSecBuff.cbBuffer, 0);
+        }
+
+    }
+    catch(...)

```

Comment:
>catch(...) [](start = 4, length = 10)

end your catch with a throw; statement? you are swallowing all exception without a trace here on purpose?

Reply:
yes. This is a delegate function called from the network stack. No exceptions shoul;d escape here. We are setting the error in a member variable:  
pSock->m_SecurityStatus = winrt::to_hresult();

This is checked at line 170, and exception is thrown into the appropriate call stack.

---
In reply to: [492999125](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r492999125) [](ancestors = 492999125)